### PR TITLE
wip add wesnoth.create_side

### PR DESCRIPTION
**Pull request by gfgtdf**

### CODE REVIEW

**Comment by CelticMinstrel** [comment]
Looks like a good start! A few comments on what needs to be done still...

* Update the `wesnoth.sides` array after adding the new side (unless we make it an array with a metatable that automatically looks up a side)
* Take a WML object defining the side
* I tried creating a side in the Lua console, and though it seemed to work at first, the game crashed when I closed the console. When I tried doing the same with the `:lua` command, it crashed immediately. Here's the stack trace in case it's helpful:

```
wesnoth.exe!boost::multi_index::multi_index_container<std::shared_ptr<wb::action>,boost::multi_index::indexed_by<boost::multi_index::random_access<boost::multi_index::tag<wb::side_actions_container::chronological,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na> >,boost::multi_index::hashed_non_unique<boost::multi_index::tag<wb::side_actions_container::by_unit,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na>,boost::multi_index::const_mem_fun<wb::action,unsigned int,{wb::action::`vcall'{44}',0}>,boost::mpl::na,boost::mpl::na>,boost::multi_index::hashed_non_unique<boost::multi_index::tag<wb::side_actions_container::by_hex,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na>,boost::multi_index::const_mem_fun<wb::action,map_location,{wb::action::`vcall'{32}',0}>,boost::mpl::na,boost::mpl::na>,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na>,std::allocator<std::shared_ptr<wb::action> > >::empty_() Line 541 (q:\Projects\Battle_for_Wesnoth\external\include\boost\multi_index_container.hpp:541)
wesnoth.exe!boost::multi_index::detail::index_base<std::shared_ptr<wb::action>,boost::multi_index::indexed_by<boost::multi_index::random_access<boost::multi_index::tag<wb::side_actions_container::chronological,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na> >,boost::multi_index::hashed_non_unique<boost::multi_index::tag<wb::side_actions_container::by_unit,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na>,boost::multi_index::const_mem_fun<wb::action,unsigned int,{wb::action::`vcall'{44}',0}>,boost::mpl::na,boost::mpl::na>,boost::multi_index::hashed_non_unique<boost::multi_index::tag<wb::side_actions_container::by_hex,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na>,boost::multi_index::const_mem_fun<wb::action,map_location,{wb::action::`vcall'{32}',0}>,boost::mpl::na,boost::mpl::na>,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na>,std::allocator<std::shared_ptr<wb::action> > >::final_empty_() Line 214 (q:\Projects\Battle_for_Wesnoth\external\include\boost\multi_index\detail\index_base.hpp:214)
wesnoth.exe!boost::multi_index::detail::random_access_index<boost::multi_index::detail::nth_layer<1,std::shared_ptr<wb::action>,boost::multi_index::indexed_by<boost::multi_index::random_access<boost::multi_index::tag<wb::side_actions_container::chronological,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na> >,boost::multi_index::hashed_non_unique<boost::multi_index::tag<wb::side_actions_container::by_unit,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na>,boost::multi_index::const_mem_fun<wb::action,unsigned int,{wb::action::`vcall'{44}',0}>,boost::mpl::na,boost::mpl::na>,boost::multi_index::hashed_non_unique<boost::multi_index::tag<wb::side_actions_container::by_hex,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na>,boost::multi_index::const_mem_fun<wb::action,map_location,{wb::action::`vcall'{32}',0}>,boost::mpl::na,boost::mpl::na>,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na,boost::mpl::na>,std::allocator<std::shared_ptr<wb::action> > >,boost::mpl::vector1<wb::side_actions_container::chronological> >::empty() Line 262 (q:\Projects\Battle_for_Wesnoth\external\include\boost\multi_index\random_access_index.hpp:262)
wesnoth.exe!wb::side_actions_container::empty() Line 197 (q:\Projects\Battle_for_Wesnoth\wesnoth\src\whiteboard\side_actions.hpp:197)
wesnoth.exe!wb::side_actions::empty() Line 326 (q:\Projects\Battle_for_Wesnoth\wesnoth\src\whiteboard\side_actions.hpp:326)
wesnoth.exe!wb::has_actions() Line 173 (q:\Projects\Battle_for_Wesnoth\wesnoth\src\whiteboard\utility.cpp:173)
wesnoth.exe!wb::manager::has_actions() Line 1085 (q:\Projects\Battle_for_Wesnoth\wesnoth\src\whiteboard\manager.cpp:1085)
wesnoth.exe!wb::manager::pre_draw() Line 541 (q:\Projects\Battle_for_Wesnoth\wesnoth\src\whiteboard\manager.cpp:541)
wesnoth.exe!game_display::pre_draw() Line 212 (q:\Projects\Battle_for_Wesnoth\wesnoth\src\game_display.cpp:212)
wesnoth.exe!display::draw(bool update, bool force) Line 2493 (q:\Projects\Battle_for_Wesnoth\wesnoth\src\display.cpp:2493)
wesnoth.exe!display::draw() Line 2461 (q:\Projects\Battle_for_Wesnoth\wesnoth\src\display.cpp:2461)
wesnoth.exe!events::raise_draw_event() Line 784 (q:\Projects\Battle_for_Wesnoth\wesnoth\src\events.cpp:784)
wesnoth.exe!controller_base::play_slice(bool is_delay_enabled) Line 358 (q:\Projects\Battle_for_Wesnoth\wesnoth\src\controller_base.cpp:358)
wesnoth.exe!play_controller::play_slice_catch() Line 1097 (q:\Projects\Battle_for_Wesnoth\wesnoth\src\play_controller.cpp:1097)
wesnoth.exe!playsingle_controller::play_human_turn() Line 474 (q:\Projects\Battle_for_Wesnoth\wesnoth\src\playsingle_controller.cpp:474)
wesnoth.exe!playsingle_controller::play_side_impl() Line 393 (q:\Projects\Battle_for_Wesnoth\wesnoth\src\playsingle_controller.cpp:393)
wesnoth.exe!play_controller::play_side() Line 1189 (q:\Projects\Battle_for_Wesnoth\wesnoth\src\play_controller.cpp:1189)
wesnoth.exe!play_controller::play_turn() Line 1230 (q:\Projects\Battle_for_Wesnoth\wesnoth\src\play_controller.cpp:1230)
wesnoth.exe!playsingle_controller::play_scenario_main_loop() Line 175 (q:\Projects\Battle_for_Wesnoth\wesnoth\src\playsingle_controller.cpp:175)
wesnoth.exe!playsingle_controller::play_scenario(const config & level) Line 273 (q:\Projects\Battle_for_Wesnoth\wesnoth\src\playsingle_controller.cpp:273)
```

**Comment by gfgtdf** [comment]
okay thank will look into that

**Comment by gfgtdf** [comment]
> I tried creating a side in the Lua console, and though it seemed

this os probably team::planned_actions_ being a nullptr, shoudl be fixed by calling team::build

**Comment by gfgtdf** [comment]
speaking of team::build is ti known what happens when two sides claim to own a village in heir [team] definition?

**Comment by CelticMinstrel** [comment]
Hey uh, sorry, I kinda rendered part of your second commit redundant in 212a91113c49b0078f0713db301d4090b540fff7

**Comment by gfgtdf** [comment]
> Basically wesnoth.sides is no longer just a static table, but instead has a metatable that looks up the side on the fly.

Does iterating over wesnoth.sides still work?

**Comment by CelticMinstrel** [comment]
Iterating with `ipairs` or a straight for loop should work (because I added the `__len` metamethod), but I think `pairs` would not work. That could be easily changed, if needed.

File: src/scripting/game_lua_kernel.cpp [hunk 1]
```diff
@@ -773,6 +773,16 @@ int game_lua_kernel::intf_set_variable(lua_State *L)
 	return 0;
 }
 
+
+int game_lua_kernel::intf_create_side(lua_State *L)
+{
+	int new_side_num = board().teams().size() + 1;
```
Comment (stevecotton) [inline src/scripting/game_lua_kernel.cpp]: `new_side_num` is unused

**Pull request merged**